Initial setup of AblyD

tomczoink opened the pull request:
Demo which allows one to wrap up any terminal process on a device and make it executable and have its stdin/stdout available over Ably. This extends the websocketd functionality to make use of Ably rather than a WebSocket server hosted on the main device/server.

Main checks are to state what needs further clarification in the README/notes, and code improvements recommendations.

tomczoink committed:
Add basic AblyD layout

tomczoink committed:
Remove Action message element

Instead of a message element to define the action, we instead specify this with the message name.

tomczoink committed:
Add basic AsyncAPI spec

This currently only defines normal messages, not presence messages.

tomczoink committed:
Fix ably key flag in README

Incorrectly said the flag for the Ably API key was `ablykey` rather than `apikey`

tomczoink committed:
Adjust logging to be per-process

Means PID in logging can identify each individual process

tomczoink committed:
Add executable

tomczoink committed:
Clean up sample count html file

tomczoink committed:
Remove misleading usage info from the count.html example

Jmgr commented:
I don't think I will be able to have a look at this today because of support queries and investigations. Could you have a look @cruickshankpg or @lmars please?

tomczoink committed:
Update channel names to be more informative

Rather than relying on a single level of channel namespace based on PID, this reflects the serverID and a variable base namespace to allow for further customization.

This includes improvements to allow for commands to be used on specific servers rather than any server in a shared command channel.

tomczoink committed:
Convert process output from bytes to string type

This should make it easier for subscribers to handle output from servers.

tomczoink committed:
Add webhook demo

tomczoink committed:
Update usage of instances to processes

This is to help differentiate between AblyD instances and the processes it spools off.

tomczoink committed:
Add note on ablyd client in NodeJS

tomczoink committed:
Update ablyd binary

lmars commented on README.md:
````diff
@@ -0,0 +1,65 @@
+Now all you need to do is run the main go file, passing in the command line command you want to be run as a parameter! For example, to use a bash file `count.sh`, which is included in this repo's `examples` folder, just run:
+
+```bash
+~ $ ./ablyD --apikey=YOUR_API_KEY ./examples/bash/count.sh
````
I recommend expecting the API key to be passed as an environment variable rather than a command line parameter, since command line parameters leak more easily.

lmars commented on README.md:
````diff
@@ -0,0 +1,65 @@
+~ $ ./ablyD --apikey=YOUR_API_KEY ./examples/bash/count.sh
+```
+
+The program is now running, waiting for a message in the `command` channel in Ably to be sent. The message's data field should match the structure, with the value `start` for the message's name:
````
I sent a message to the `command` channel and nothing happened, after digging a little bit it seems the channel needs to be `ablyd:command`.

I'd recommend giving a full example of the message that should be published, for example using cURL:

```sh
curl -X POST https://rest.ably.io/channels/ablyd:command/messages \
  -u "${API_KEY}" \
  -H 'Content-Type: application/json' \
  --data \
  '{
    "name": "start",
    "data": {
      "MessageID": "unique string value",
      "Args": [ "some", "additional", "args", "for", "the", "programs" ]
    },
    "format":"json"
  }'
```

lmars commented on README.md:
````diff
@@ -0,0 +1,65 @@
+}
+```
+
+The client can identify the process which has started for them by the `MessageID`, then use the `Prefix` to connect to an input and an output channel for the process. These will be of structure `{Prefix}{Pid}:serverinput` and `{Prefix}{Pid}:serveroutput`.
````
It would be useful if the command message included the names of the channels explicitly, for example:

```json
{
  "MessageID": "unique string value matching the requesting MessageID",
  "Stdin": "ably:c42jl786n88h66cnsu70:5328:serverinput",
  "Stdout": "ably:c42jl786n88h66cnsu70:5328:serveroutput"
}
```

lmars commented on README.md:
```diff
@@ -0,0 +1,65 @@
+
+The client can identify the process which has started for them by the `MessageID`, then use the `Prefix` to connect to an input and an output channel for the process. These will be of structure `{Prefix}{Pid}:serverinput` and `{Prefix}{Pid}:serveroutput`.
+
+Subscribing to the `serveroutput` channel will allow the client to receive any stdout messages from the server. The client can also publish messages into the `serverinput` channel which will be passed into the stdin of the process.
```
This design is a little racy, since I only get to know the channels to subscribe to after the process has started, and so in my testing I sometimes miss the first bit of output.

It would be better if I could create the command, get back an identifier, subscribe to the channels, and then start the command, this way I'll always be subscribed before anything gets published to the output channel.

I guess this would need ablyD to generate random identifiers for commands rather than using the PID (since the PID won't be known before starting the command), but I don't think that presents a problem.

lmars commented on README.md:
```diff
@@ -0,0 +1,65 @@
+
+Subscribing to the `serveroutput` channel will allow the client to receive any stdout messages from the server. The client can also publish messages into the `serverinput` channel which will be passed into the stdin of the process.
+
+This will continue until the program naturally terminates, resulting in the process dying, or the client submits a message to the `serverinput` with data `KILL`.
```
We should consider adding a third `signal` channel the client can use to send OS signals to the process rather than trying to re-purpose the stdin channel.

That being said, we should also consider just using a single channel with typed messages, rather than using different channels for stdin / stdout / signals. For example, there could be a single "data" channel with messages like `{"type":"STDIN","data":"..."}` and `{"type":"SIGNAL","data":"15"}`. This is pretty much what the SSH protocol does (see https://datatracker.ietf.org/doc/html/rfc4254#section-5), but is probably out of scope for what you're trying to achieve here 🙂

tomczoink committed:
Switch to using API key from env

tomczoink committed:
Add goreleaser

tomczoink committed:
Update homebrew url

tomczoink committed:
Add link to blog post

tomczoink committed:
Add  Install details
